src/object/srv_obj.c

```diff
@@ -2313,7 +2314,7 @@ obj_inflight_io_check(struct ds_cont_child *child, uint32_t opc,
 		}
 	}
 
-	if (!obj_is_modification_opc(opc))
+	if (!obj_is_modification_opc(opc) && (opc != DAOS_OBJ_RPC_CPD || flags & ORF_CPD_RDONLY))
 		return 0;
```

Comment:
just confirm for CPD request, will it carry ORF_REINTEGRATING_IO/ORF_REBUILDING_IO flag which will be checked in following code section?
just checked the code seems oci_flags only carry ORF_REINTEGRATING_IO, no where carry ORF_REBUILDING_IO flag found. Not very sure how does it work for that checking?

Reply:
The flag ORF_REINTEGRATING_IO for CPD RPC was introduced via the commit 7010f2eb822f8b81547386230d7e84620c8f0c74. As my understand, CPR RPC with such flag only happens during reintegration caused rebuild instead of rebuild for target/rank exclusion. As the user reported, the new written data was lost because of discarding old data when rebuild, that is for reintegration. It is expected to be avoided via the following logic:

`2332 
2333         /* If the incoming I/O is during integration, then it needs to wait the
2334          * vos discard to finish, which otherwise might discard these new in-flight
2335          * I/O update.
2336          */
2337         if ((flags & ORF_REINTEGRATING_IO) &&
2338             (child->sc_pool->spc_pool->sp_need_discard &&
2339              child->sc_pool->spc_discard_done == 0)) {
2340                 D_ERROR("reintegrating "DF_UUID" retry.\n", DP_UUID(child->sc_pool->spc_uuid));
2341                 return -DER_UPDATE_AGAIN;
2342         }
2343 `

If without this patch, CPD RPC will always be regarded as non-modification operations, as to pass the obj_inflight_io_check() without returning the expected `-DER_UPDATE_AGAIN`.

Reply:
I was thought about the code of checking ORF_REBUILDING_IO flag in obj_inflight_io_check, I think that code can be removed as now vos can support overwrite. I can do it in another patch. thx